### PR TITLE
Ping for EdgeAgent

### PR DESCRIPTION
**Pull request by yophilav**
Bugfix: 
- Update DirectMethod Receiver to correctly send the test result to TRC. 

Updates: 
- Deploy DirectMethod Sender to send `ping` method to EdgeAgent

### CODE REVIEW

File: test/modules/ModuleLib/ModuleUtil.cs [hunk 1]
```diff
@@ -56,7 +56,7 @@ public static ILogger CreateLogger(string categoryName, LogEventLevel logEventLe
             return new LoggerFactory().AddSerilog().CreateLogger(categoryName);
         }
 
-        public static async Task ReportTestResultAsync(TestResultReportingClient apiClient, ILogger logger, TestResultBase testResult)
+        public static async Task ReportTestResultAsync(TestResultReportingClient apiClient, ILogger logger, TestResultBase testResult, bool isRethrow = false)
```
Comment (philipktlin) [inline test/modules/ModuleLib/ModuleUtil.cs]: Can we always call throw? and lets' caller to handle.

File: test/modules/TestResultCoordinator/Settings.cs [hunk 2]
```diff
@@ -166,6 +167,7 @@ List<string> GetResultSources()
                 "relayer2.eventHub",
                 "directMethodSender1.send",
                 "directMethodSender2.send",
+                "directMethodSender3.send",
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Settings.cs]: what are you going to expect if both actual and expected are always the same?

File: test/modules/ModuleLib/ModuleUtil.cs (hunk 1 above)
Comment (yophilav) [reply test/modules/ModuleLib/ModuleUtil.cs]: It looks like there are multiple caller to this function that I don't want to make change right now. 
So, I'll make the default behavior a swallow exception. And if a new test module want the manage the exception, the caller can set the flag to be true.

File: test/modules/TestResultCoordinator/Settings.cs (hunk 2 above)
Comment (yophilav) [reply test/modules/TestResultCoordinator/Settings.cs]: I actually have no idea. Dylan is working creating&verify process for DM tests.

File: test/modules/ModuleLib/ModuleUtil.cs (hunk 1 above)
Comment (philipktlin) [reply test/modules/ModuleLib/ModuleUtil.cs]: revert this fix.  I will take a look and better to handle in a separate PR.

File: test/modules/TestResultCoordinator/Settings.cs [hunk 3]
```diff
@@ -143,6 +143,7 @@ List<IReportMetadata> InitializeReportMetadataList()
                 new CountingReportMetadata("relayer2.send", "relayer2.eventHub", TestOperationResultType.Messages, TestReportType.CountingReport),
                 new CountingReportMetadata("directMethodSender1.send", "directMethodReceiver1.receive", TestOperationResultType.DirectMethod, TestReportType.CountingReport),
                 new CountingReportMetadata("directMethodSender2.send", "directMethodReceiver2.receive", TestOperationResultType.DirectMethod, TestReportType.CountingReport),
+                new CountingReportMetadata("directMethodSender3.send", "directMethodSender3.send", TestOperationResultType.DirectMethod, TestReportType.CountingReport),
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Settings.cs]: Remove this line for now.

File: e2e_deployment_files/connectivity_deployment.template.json [hunk 4]
```diff
@@ -222,6 +222,46 @@
...
+              "DirectMethodResultType": {
+                "value": "DirectMethodTestResult"
+              },
+              "DirectMethodName": {
+                "value": "ping"
+              },
+              "InvocationSource": {
+                "value": "Cloud"
+              },
+              "ServiceClientConnectionString": {
```
Comment (philipktlin) [inline e2e_deployment_files/connectivity_deployment.template.json]: Rename it to IOT_HUB_CONNECTION_STRING.

File: test/modules/TestResultCoordinator/Settings.cs [hunk 5]
```diff
@@ -143,6 +143,8 @@ List<IReportMetadata> InitializeReportMetadataList()
                 new CountingReportMetadata("relayer2.send", "relayer2.eventHub", TestOperationResultType.Messages, TestReportType.CountingReport),
                 new CountingReportMetadata("directMethodSender1.send", "directMethodReceiver1.receive", TestOperationResultType.DirectMethod, TestReportType.CountingReport),
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Settings.cs]: please comment out the other 2 direct method reports as well.

File: e2e_deployment_files/connectivity_deployment.template.json (hunk 4 above)
Comment (yophilav) [reply e2e_deployment_files/connectivity_deployment.template.json]: The `deploymentTest1` is the only module that uses `IOT_HUB_CONNECTION_STRING` in Connectivity Test. Would you like me to change that one instead?

File: test/modules/TestResultCoordinator/Settings.cs (hunk 5 above)
Comment (yophilav) [reply test/modules/TestResultCoordinator/Settings.cs]: yessir

File: test/modules/TestResultCoordinator/Settings.cs [hunk 6]
```diff
@@ -141,8 +141,10 @@ List<IReportMetadata> InitializeReportMetadataList()
                 new CountingReportMetadata("relayer1.send", "relayer1.eventHub", TestOperationResultType.Messages, TestReportType.CountingReport),
                 new CountingReportMetadata("loadGen2.send", "relayer2.receive", TestOperationResultType.Messages, TestReportType.CountingReport),
                 new CountingReportMetadata("relayer2.send", "relayer2.eventHub", TestOperationResultType.Messages, TestReportType.CountingReport),
-                new CountingReportMetadata("directMethodSender1.send", "directMethodReceiver1.receive", TestOperationResultType.DirectMethod, TestReportType.CountingReport),
-                new CountingReportMetadata("directMethodSender2.send", "directMethodReceiver2.receive", TestOperationResultType.DirectMethod, TestReportType.CountingReport),
+                // TODO: Enable Direct Method Cloud-to-Module and Cloud-to-EdgeAgent once the verification scheme is finallized.
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Settings.cs]: typo
```suggestion
                // TODO: Enable Direct Method Cloud-to-Module and Cloud-to-EdgeAgent once the verification scheme is finalized.
```

File: test/modules/TestResultCoordinator/Settings.cs (hunk 2 above)
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/Settings.cs]: Ping for EdgeAgent test won't need two stores - it can have only one. I think we should either make a new DirectMethodReport type for this case, or make one of the stores Optional and generate a different report depending on if the option is filled in or not.

File: e2e_deployment_files/connectivity_deployment.template.json [hunk 7]
```diff
@@ -119,7 +119,7 @@
               "InvocationSource": {
                 "value": "Cloud"
               },
-              "ServiceClientConnectionString": {
+              "IOT_HUB_CONNECTION_STRING": {
```
Comment (varunpuranik) [inline e2e_deployment_files/connectivity_deployment.template.json]: Why this change?

File: e2e_deployment_files/connectivity_deployment.template.json [hunk 8]
```diff
@@ -222,6 +222,46 @@
               "createOptions": ""
             }
           },
+          "directMethodSender3": {
```
Comment (varunpuranik) [inline e2e_deployment_files/connectivity_deployment.template.json]: Maybe we should call this edgeAgentDirectMethodSender or something like that?

File: test/modules/TestResultCoordinator/Settings.cs [hunk 9]
```diff
@@ -141,8 +141,10 @@ List<IReportMetadata> InitializeReportMetadataList()
                 new CountingReportMetadata("relayer1.send", "relayer1.eventHub", TestOperationResultType.Messages, TestReportType.CountingReport),
                 new CountingReportMetadata("loadGen2.send", "relayer2.receive", TestOperationResultType.Messages, TestReportType.CountingReport),
                 new CountingReportMetadata("relayer2.send", "relayer2.eventHub", TestOperationResultType.Messages, TestReportType.CountingReport),
-                new CountingReportMetadata("directMethodSender1.send", "directMethodReceiver1.receive", TestOperationResultType.DirectMethod, TestReportType.CountingReport),
-                new CountingReportMetadata("directMethodSender2.send", "directMethodReceiver2.receive", TestOperationResultType.DirectMethod, TestReportType.CountingReport),
+                // TODO: Enable Direct Method Cloud-to-Module and Cloud-to-EdgeAgent once the verification scheme is finalized.
```
Comment (varunpuranik) [inline test/modules/TestResultCoordinator/Settings.cs]: Can you clarify why we need to comment this out?

Comment (yophilav) [reply test/modules/TestResultCoordinator/Settings.cs]: The verification & report generating part is not there yet. 
It has a dependency on this PR: https://github.com/Azure/iotedge/pull/2300/files

File: e2e_deployment_files/connectivity_deployment.template.json (hunk 7 above)
Comment (yophilav) [reply e2e_deployment_files/connectivity_deployment.template.json]: `ServiceClientConnectionString` is a variable used in Long Haul test. According to Philip's comment (https://github.com/Azure/iotedge/pull/2286#discussion_r365924786), he would like to use the variable `IOT_HUB_CONNECTION_STRING` in connectivity test instead.

File: e2e_deployment_files/connectivity_deployment.template.json (hunk 8 above)
Comment (yophilav) [reply e2e_deployment_files/connectivity_deployment.template.json]: we could do `edgeAgentDirectMethodSender` but we decided to do the enumeration for the test module instead (at least for now).

File: test/modules/TestResultCoordinator/Settings.cs (hunk 9 above)
Comment (philipktlin) [reply test/modules/TestResultCoordinator/Settings.cs]: It is better to comment it out since these counting report should not be used for DM scenario which created noise in test report result.

File: e2e_deployment_files/connectivity_deployment.template.json (hunk 8 above)
Comment (philipktlin) [reply e2e_deployment_files/connectivity_deployment.template.json]: I think Varun's suggestion is doable, can you update in this PR?
